### ISSUE TITLE
Example - Regression

### ISSUE THREAD

**Issue by hmustafajr** [hmustafajr/Python-Basic-Concepts#2]
We are going to explore a regression with a simplified example where we'll train a model to predict a numeric label (y) based on a single feature value (x).
**Note**: Most real life scenarios involve multiple feature values, which adds complexity; but the principle is the same.

For our example, let's use a Seattle Dog sales scenario. For our feature, we'll consider the day of the week, and the label we want to train a model to predict is the number of Seattle Dogs sold that day. We'll start with some historic data that includes the days of the week (x) and Seattle dog sales (y):

<html>
<body>

Day of the week (x) | Seattle Dog sales (y)
-- | --
1 | 5
2 | 4
3 | 12
4 | 3
5 | 10
6 | 13
7 | 11

</body>
</html>

**Training a regression model**
We'll start by segmenting the data and using a subset of it to train a model. Here's the training dataset:
<html>
<body>

Day of the Week (x) | Seattle Dog sales (y)
-- | --
4 | 3
6 | 13

</body>
</html>

To get an insight of how these x and y values might relate to one another, we'll plot them as coordinates along two axes, like this:

Now we can apply an algorithm to our training data and fit it to a function that applies an operation to x to calculate y. This will be the linear expression algorithm, which works by deriving a function that produces a straight line through the intersections of the x and y values while minimizing the average distance between the line and the plotted points, like the image below:

![Image](https://github.com/hmustafajr/Python-Basic-Concepts/assets/47394208/60409c1b-54a3-4d42-869a-22fa1061b967)

The line is a visual representation of the function in which the slope of the line describes how to calculate the value of y for a given value of x. The line intercepts the x axis at 1, so when x is 1, y is 5. Meaning the week starts on Monday (1) and the Seattle Dogs sold was (5).

As you can see from the axis markers in the plot, the line slopes so that later in the week (every increase of 1 along the x axis) results in an increase of 4.5 up the y axis (increase in Seattle Dogs sold); so when x is 1, y is 5; when x is 5, y is 10, and so on. In other words, when it is Monday the number of dogs sold is likely 5, when it is Friday the number of dogs sold is likely 10.

To calculate a value of y for a given value of x, the function simply adds 4.552; in other words, the function can be expressed like this:

f(x) = x +4.552

You can use this function to predict the number of Seattle Dogs sold in a day with any given day of the week. For example, suppose it's Saturday. We can apply our model to calculate 6 + 4.552 and predict that we'll probably sell 11 Seattle Dogs on Saturday.

But just how accurate is our model?

**Evaluating a regression model**

To validate the model and evaluate how well it predicts, we held back some data for which we know the label (y) value. Here's the data we held back:

<html>
<body>

Day of the Week (x) | Seattle Dog sales (y)
-- | --
4 | 3
6 | 13

</body>
</html>

We can use the model to predict the label for each of the observations in this dataset based on the feature (x) value; and then compare the predicted label (ŷ) to the known actual label value (y).

Using the model we trained earlier, which encapsulates the function f(x) = x+4.552, results in the following predictions:

![Image](https://github.com/hmustafajr/Python-Basic-Concepts/assets/47394208/231a96e2-46f0-463f-8b79-6532d2a0df24)

The predicted labels are calculated by the model so they're on the function line, but there's some variance between the ŷ values calculated by the function and the actual y values from the validation dataset; which is indicated on the plot as a line between the ŷ and y values that shows how far off the prediction was from the actual value.

**Regression evaluation metrics**
Based on the differences between the predicted and actual values, you can calculate some common metrics that are used to evaluate a regression model.

**Mean Absolute Error (MAE)**
The variance in this example indicates by how many Seattle Dogs each prediction was not accurate or wrong. Regardless if the prediction was over or under the actual value, This metric is known as the absolute error for each prediction, and can be summarized for the entire validation set as the Mean Absolute Error (MAE).

In summary, the MAE is the average of the absolute errors (v1 + v2 + v3 + v4 + v5 + v6 + v7) / 7. Where v represents the variance.

**Mean Squared Error (MSE)**
The mean absolute error metric takes all discrepancies between predicted and actual labels into account equally. However, it may be more desirable to have a model that is consistently wrong by a small amount than one that makes fewer, but larger errors.

One way to produce a metric that "amplifies" larger errors by squaring the individual errors then calculating the mean of the squared values, This metric is known as the Mean Squared Error (MSE).

In our Seattle Dog example, the mean of the squared absolute values which are the square of each variance.

**Root Mean Squared Error (RMSE)**
The mean squared error helps take the magnitude of errors into account, because it squares the error values, the resulting metric no longer represents the quantity measured by the label (y). For clarity, we can say that the MSE of our model is MSE, but that doesn't measure for accuracy in terms of the number of Seattle Dogs that were mispredicted. The MSE is pretty much just a number score that indicates the level of error in the validation predictions.

If we want to measure the error in terms of the number of Seattle Dogs, we should calculate the sq root of the MSE - which produces a metric called, unsurprisingly, Root Mean Squared Error. 

**Iterative training**
The metrics for Coefficent of determination (R2) (coming soon) will be used to evaluate a regression model. In most real-world scenarios, a data scientist will use an iterative process to repeatedly train and evaluate a model, varying:

- Feature selection and preparation (choosing which features to include in the model, and calculations applied to them to help ensure a better fit).
- Algorithm selection (We explored linear regression but there are many other regression algorithms).
- Algorithm parameters (numeric settings that control algorithm behavior, more accurately called hyperparameters to differentiate them from the x and y markers).

After multiple iterations, the model that results in the best evaluation metric that's acceptable for the specific scenario is chosen.

**Comment by hmustafajr** [comment]
Need to figure out how to add Coefficient of determination (R2)
